File: shared/prompt-schema.js
(function (global) {
    'use strict';

    const REQUIRED_STRING_FIELDS = ['id', 'title', 'category', 'text'];

    function isPlainObject(value) {
        return value !== null && typeof value === 'object' && !Array.isArray(value);
    }

    function normalizePrompt(record) {
        const errors = [];

        if (!isPlainObject(record)) {
            return { prompt: null, errors: ['Prompt must be an object.'] };
        }

        const prompt = {
            id: typeof record.id === 'string' ? record.id.trim() : '',
            title: typeof record.title === 'string' ? record.title.trim() : '',
            category: typeof record.category === 'string' ? record.category.trim() : '',
            description: typeof record.description === 'string' ? record.description.trim() : '',
            text: typeof record.text === 'string' ? record.text.trim() : '',
            pinned: record.pinned === true
        };

        REQUIRED_STRING_FIELDS.forEach(field => {
            if (!prompt[field]) {
                errors.push(`Missing or invalid required field: ${field}.`);
            }
        });

        if (record.description !== undefined && typeof record.description !== 'string') {
            errors.push('Invalid optional field: description must be a string.');
        }

        if (record.pinned !== undefined && typeof record.pinned !== 'boolean') {
            errors.push('Invalid optional field: pinned must be a boolean.');
        }

        return { prompt: errors.length ? null : prompt, errors };
    }

    function normalizePromptCollection(records) {
        if (!Array.isArray(records)) {
            return {
                prompts: [],
                issues: [{ index: null, id: null, errors: ['Prompt collection must be an array.'] }]
            };
        }

        const prompts = [];
        const issues = [];
        const seenIds = new Set();

        records.forEach((record, index) => {
            const result = normalizePrompt(record);
            const candidateId = isPlainObject(record) && typeof record.id === 'string'
                ? record.id.trim()
                : null;

            if (!result.prompt) {
                issues.push({ index, id: candidateId, errors: result.errors });
                return;
            }

            if (seenIds.has(result.prompt.id)) {
                issues.push({
                    index,
                    id: result.prompt.id,
                    errors: [`Duplicate prompt ID: ${result.prompt.id}.`]
                });
                return;
            }

            seenIds.add(result.prompt.id);
            prompts.push(result.prompt);
        });

        return { prompts, issues };
    }

    global.PromptSchema = Object.freeze({
        normalizePrompt,
        normalizePromptCollection
    });
})(window);

File: launcher/PromptLauncher.Core/PromptSchema.cs
using System.Text.Json;

namespace PromptLauncher.Core;

public static class PromptSchema
{
    public static ValidationResult Normalize(JsonElement records)
    {
        if (records.ValueKind != JsonValueKind.Array)
            return new([], [new(null, null, ["Prompt collection must be an array."])]);

        var prompts = new List<PromptRecord>();
        var issues = new List<ValidationIssue>();
        var seenIds = new HashSet<string>(StringComparer.Ordinal);
        var index = 0;

        foreach (var record in records.EnumerateArray())
        {
            var (prompt, errors, candidateId) = NormalizeRecord(record);
            if (prompt is null)
            {
                issues.Add(new(index++, candidateId, errors));
                continue;
            }

            if (!seenIds.Add(prompt.Id))
            {
                issues.Add(new(index++, prompt.Id, [$"Duplicate prompt ID: {prompt.Id}."]));
                continue;
            }

            prompts.Add(prompt);
            index++;
        }

        return new(prompts, issues);
    }

    public static ValidationResult Normalize(IEnumerable<PromptRecord> records)
    {
        var json = JsonSerializer.SerializeToElement(records.Select(prompt => new
        {
            id = prompt.Id,
            title = prompt.Title,
            category = prompt.Category,
            description = prompt.Description,
            text = prompt.Text,
            pinned = prompt.Pinned
        }));
        return Normalize(json);
    }

    private static (PromptRecord? Prompt, List<string> Errors, string? CandidateId) NormalizeRecord(JsonElement record)
    {
        if (record.ValueKind != JsonValueKind.Object)
            return (null, ["Prompt must be an object."], null);

        var errors = new List<string>();
        var id = ReadString(record, "id", true, errors);
        var title = ReadString(record, "title", true, errors);
        var category = ReadString(record, "category", true, errors);
        var description = ReadString(record, "description", false, errors);
        var text = ReadString(record, "text", true, errors);
        var pinned = false;

        if (record.TryGetProperty("pinned", out var pin))
        {
            if (pin.ValueKind is JsonValueKind.True or JsonValueKind.False) pinned = pin.GetBoolean();
            else errors.Add("Invalid optional field: pinned must be a boolean.");
        }

        if (PromptTemplate.Analyze(text).EmptyPlaceholders.Count > 0)
            errors.Add("Variable names cannot be empty.");

        return errors.Count > 0
            ? (null, errors, id.Length > 0 ? id : null)
            : (new(id, title, category, description, text, pinned), errors, id);
    }

    private static string ReadString(JsonElement record, string name, bool required, List<string> errors)
    {
        if (!record.TryGetProperty(name, out var value))
        {
            if (required) errors.Add($"Missing or invalid required field: {name}.");
            return "";
        }

        if (value.ValueKind != JsonValueKind.String)
        {
            errors.Add(required
                ? $"Missing or invalid required field: {name}."
                : $"Invalid optional field: {name} must be a string.");
            return "";
        }

        var result = value.GetString()!.Trim();
        if (required && result.Length == 0)
            errors.Add($"Missing or invalid required field: {name}.");
        return result;
    }
}
